In [3]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np

from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split, ShuffleSplit, GridSearchCV
from sklearn.metrics import roc_auc_score, roc_curve, auc

import os
import time
import sys

In [4]:
df = pd.read_csv("Train data values.csv")

df2 = pd.read_csv("Train data targets or labels.csv")

In [5]:
df2.head()

id    status_group
0  69572      functional
1   8776      functional
2  34310      functional
3  67743  non functional
4  19728      functional

In [6]:
df.insert(0, "status_group", df2["status_group"], allow_duplicates = False)

In [7]:
df["bin_status"] = df.apply(lambda r: 1 if r.status_group > "functional" else 0 , axis=1)

In [8]:
df.head()

status_group     id  amount_tsh date_recorded        funder  gps_height  \
0      functional  69572      6000.0    2011-03-14         Roman        1390   
1      functional   8776         0.0    2013-03-06       Grumeti        1399   
2      functional  34310        25.0    2013-02-25  Lottery Club         686   
3  non functional  67743         0.0    2013-01-28        Unicef         263   
4      functional  19728         0.0    2011-07-13   Action In A           0   

      installer  longitude   latitude              wpt_name    ...      \
0         Roman  34.938093  -9.856322                  none    ...       
1       GRUMETI  34.698766  -2.147466              Zahanati    ...       
2  World vision  37.460664  -3.821329           Kwa Mahundi    ...       
3        UNICEF  38.486161 -11.155298  Zahanati Ya Nanyumbu    ...       
4       Artisan  31.130847  -1.825359               Shuleni    ...       

   water_quality quality_group      quantity quantity_group  \
0           soft          good        enough         enough   
1           soft          good  insufficient   insufficient   
2           soft          good        enough         enough   
3           soft          good           dry            dry   
4           soft          good      seasonal       seasonal   

                 source           source_type source_class  \
0                spring                spring  groundwater   
1  rainwater harvesting  rainwater harvesting      surface   
2                   dam                   dam      surface   
3           machine dbh              borehole  groundwater   
4  rainwater harvesting  rainwater harvesting      surface   

               waterpoint_type  waterpoint_type_group bin_status  
0           communal standpipe     communal standpipe          0  
1           communal standpipe     communal standpipe          0  
2  communal standpipe multiple     communal standpipe          0  
3  communal standpipe multiple     communal standpipe          1  
4           communal standpipe     communal standpipe          0  

[5 rows x 42 columns]

In [9]:
y=df["bin_status"]

In [10]:
df.describe()

id     amount_tsh    gps_height     longitude      latitude  \
count  59400.000000   59400.000000  59400.000000  59400.000000  5.940000e+04   
mean   37115.131768     317.650385    668.297239     34.077427 -5.706033e+00   
std    21453.128371    2997.574558    693.116350      6.567432  2.946019e+00   
min        0.000000       0.000000    -90.000000      0.000000 -1.164944e+01   
25%    18519.750000       0.000000      0.000000     33.090347 -8.540621e+00   
50%    37061.500000       0.000000    369.000000     34.908743 -5.021597e+00   
75%    55656.500000      20.000000   1319.250000     37.178387 -3.326156e+00   
max    74247.000000  350000.000000   2770.000000     40.345193 -2.000000e-08   

        num_private   region_code  district_code    population  \
count  59400.000000  59400.000000   59400.000000  59400.000000   
mean       0.474141     15.297003       5.629747    179.909983   
std       12.236230     17.587406       9.633649    471.482176   
min        0.000000      1.000000       0.000000      0.000000   
25%        0.000000      5.000000       2.000000      0.000000   
50%        0.000000     12.000000       3.000000     25.000000   
75%        0.000000     17.000000       5.000000    215.000000   
max     1776.000000     99.000000      80.000000  30500.000000   

       construction_year    bin_status  
count       59400.000000  59400.000000  
mean         1300.652475      0.456919  
std           951.620547      0.498145  
min             0.000000      0.000000  
25%             0.000000      0.000000  
50%          1986.000000      0.000000  
75%          2004.000000      1.000000  
max          2013.000000      1.000000

In [11]:
df.dtypes

status_group              object
id                         int64
amount_tsh               float64
date_recorded             object
funder                    object
gps_height                 int64
installer                 object
longitude                float64
latitude                 float64
wpt_name                  object
num_private                int64
basin                     object
subvillage                object
region                    object
region_code                int64
district_code              int64
lga                       object
ward                      object
population                 int64
public_meeting            object
recorded_by               object
scheme_management         object
scheme_name               object
permit                    object
construction_year          int64
extraction_type           object
extraction_type_group     object
extraction_type_class     object
management                object
management_group          object
payment   

In [12]:
len(df.columns)

42

In [13]:
del df["num_private"]

In [14]:
del df["id"]

In [15]:
del df["wpt_name"]

In [16]:
del df["public_meeting"]

In [17]:
del df["recorded_by"]

In [18]:
del df["scheme_name"]

In [19]:
del df["extraction_type"]

In [20]:
del df["extraction_type_group"]

In [21]:
del df["management"]

In [22]:
del df["payment"]

In [23]:
del df["quality_group"]

In [24]:
del df["source"]

In [25]:
del df["waterpoint_type"]

In [26]:
del df["lga"]

In [27]:
del df["date_recorded"]

In [28]:
del df["region_code"]

In [29]:
del df["district_code"]

In [30]:
del df["quantity_group"]

In [31]:
del df["status_group"]

In [32]:
len(df.columns)

23

In [33]:
df.columns

Index(['amount_tsh', 'funder', 'gps_height', 'installer', 'longitude',
       'latitude', 'basin', 'subvillage', 'region', 'ward', 'population',
       'scheme_management', 'permit', 'construction_year',
       'extraction_type_class', 'management_group', 'payment_type',
       'water_quality', 'quantity', 'source_type', 'source_class',
       'waterpoint_type_group', 'bin_status'],
      dtype='object')

In [34]:
df.columns.sort_values()

Index(['amount_tsh', 'basin', 'bin_status', 'construction_year',
       'extraction_type_class', 'funder', 'gps_height', 'installer',
       'latitude', 'longitude', 'management_group', 'payment_type', 'permit',
       'population', 'quantity', 'region', 'scheme_management', 'source_class',
       'source_type', 'subvillage', 'ward', 'water_quality',
       'waterpoint_type_group'],
      dtype='object')

In [35]:
[df.dtypes.sort_values()]

[bin_status                 int64
 construction_year          int64
 gps_height                 int64
 population                 int64
 latitude                 float64
 amount_tsh               float64
 longitude                float64
 basin                     object
 source_class              object
 source_type               object
 quantity                  object
 water_quality             object
 payment_type              object
 management_group          object
 extraction_type_class     object
 permit                    object
 waterpoint_type_group     object
 installer                 object
 ward                      object
 region                    object
 subvillage                object
 funder                    object
 scheme_management         object
 dtype: object]

In [36]:
cat_cols=['source_class',             
'source_type',              
'quantity',                  
'water_quality',             
'payment_type',              
'management_group',          
'extraction_type_class',     
'scheme_management',  
'waterpoint_type_group',     
'ward',                      
'region',                    
'subvillage',                
'basin',                     
'installer',                 
'funder',                    
'permit',                    
'bin_status' ]            

In [37]:
cat_pd=df[cat_cols]

In [38]:
for each in cat_cols:
    print(each,cat_pd[each].nunique())

source_class 3
source_type 7
quantity 5
water_quality 8
payment_type 7
management_group 5
extraction_type_class 7
scheme_management 12
waterpoint_type_group 6
ward 2092
region 21
subvillage 19287
basin 9
installer 2145
funder 1897
permit 2
bin_status 2


# gotta drop these or label encode
* ward 2092
* subvillage 19287
* installer 2145
* funder 1897



In [39]:
df.shape

(59400, 23)

In [40]:
features_to_use = [
    'amount_tsh', 'gps_height',
       'longitude', 'latitude', 'basin', 'region',
       'population', 'scheme_management', 'permit',
       'construction_year', 'extraction_type_class', 'management_group',
       'payment_type', 'water_quality', 'quantity', 'source_type',
       'source_class', 'waterpoint_type_group'
]

In [41]:
X = df[features_to_use]

In [42]:
y = df["bin_status"]

In [43]:
X.head()

amount_tsh  gps_height  longitude   latitude                    basin  \
0      6000.0        1390  34.938093  -9.856322               Lake Nyasa   
1         0.0        1399  34.698766  -2.147466            Lake Victoria   
2        25.0         686  37.460664  -3.821329                  Pangani   
3         0.0         263  38.486161 -11.155298  Ruvuma / Southern Coast   
4         0.0           0  31.130847  -1.825359            Lake Victoria   

    region  population scheme_management permit  construction_year  \
0   Iringa         109               VWC  False               1999   
1     Mara         280             Other   True               2010   
2  Manyara         250               VWC   True               2009   
3   Mtwara          58               VWC   True               1986   
4   Kagera           0               NaN   True                  0   

  extraction_type_class management_group payment_type water_quality  \
0               gravity       user-group     annually          soft   
1               gravity       user-group    never pay          soft   
2               gravity       user-group   per bucket          soft   
3           submersible       user-group    never pay          soft   
4               gravity            other    never pay          soft   

       quantity           source_type source_class waterpoint_type_group  
0        enough                spring  groundwater    communal standpipe  
1  insufficient  rainwater harvesting      surface    communal standpipe  
2        enough                   dam      surface    communal standpipe  
3           dry              borehole  groundwater    communal standpipe  
4      seasonal  rainwater harvesting      surface    communal standpipe

In [44]:
y.head()

0    0
1    0
2    0
3    1
4    0
Name: bin_status, dtype: int64

In [45]:
X_encoded = pd.get_dummies(X)

In [46]:
X_train, X_test, y_train, y_test = train_test_split(X_encoded, y, random_state=42)

In [47]:
X_train.shape, X_test.shape

((44550, 98), (14850, 98))

In [48]:
X_train.head()

amount_tsh  gps_height  longitude      latitude  population  \
24947        20.0         330  38.123839 -6.087137e+00          80   
22630         0.0           0   0.000000 -2.000000e-08           0   
13789         0.0           0  33.312321 -2.814100e+00           0   
15697         0.0        1542  34.783049 -4.842093e+00          23   
22613         0.0         523  34.660944 -1.070733e+01           1   

       construction_year  basin_Internal  basin_Lake Nyasa  basin_Lake Rukwa  \
24947               1972               0                 0                 0   
22630                  0               0                 0                 0   
13789                  0               0                 0                 0   
15697               2010               1                 0                 0   
22613               2013               0                 1                 0   

       basin_Lake Tanganyika             ...               source_type_spring  \
24947                      0             ...                                0   
22630                      0             ...                                0   
13789                      0             ...                                0   
15697                      0             ...                                0   
22613                      0             ...                                1   

       source_class_groundwater  source_class_surface  source_class_unknown  \
24947                         1                     0                     0   
22630                         1                     0                     0   
13789                         1                     0                     0   
15697                         0                     1                     0   
22613                         1                     0                     0   

       waterpoint_type_group_cattle trough  \
24947                                    0   
22630                                    0   
13789                                    0   
15697                                    0   
22613                                    0   

       waterpoint_type_group_communal standpipe  waterpoint_type_group_dam  \
24947                                         1                          0   
22630                                         0                          0   
13789                                         0                          0   
15697                                         1                          0   
22613                                         1                          0   

       waterpoint_type_group_hand pump  waterpoint_type_group_improved spring  \
24947                                0                                      0   
22630                                1                                      0   
13789                                1                                      0   
15697                                0                                      0   
22613                                0                                      0   

       waterpoint_type_group_other  
24947                            0  
22630                            0  
13789                            0  
15697                            0  
22613                            0  

[5 rows x 98 columns]

In [49]:
sds = StandardScaler()
sds.fit(X_train)
X_train = sds.transform(X_train)
X_test = sds.transform(X_test)

In [50]:
RandomForestClassifier(n_estimators=100,max_leaf_nodes=12,max_depth=30)

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=30, max_features='auto', max_leaf_nodes=12,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=100, n_jobs=1,
            oob_score=False, random_state=None, verbose=0,
            warm_start=False)

In [51]:
t0 = time.time()
clf = RandomForestClassifier(n_estimators=60,n_jobs=12)
  
clf.fit(X_train, y_train)
t1 = time.time()
   
ypred = clf.predict(X_test)
scores = roc_auc_score(y_test, ypred)
    
print(f"Random Forest took {t1-t0} seconds to train and got a score of {scores}")
print()
print("And here are the top features: ")
c = X_encoded.columns 
for cfi in zip(c, clf.feature_importances_):
    print(cfi)
print()

Random Forest took 1.5350191593170166 seconds to train and got a score of 0.8078197003765483

And here are the top features: 
('amount_tsh', 0.03002392418478126)
('gps_height', 0.07664145658095617)
('longitude', 0.16072056446263924)
('latitude', 0.15763310775027048)
('population', 0.05266268129545824)
('construction_year', 0.060753450265881845)
('basin_Internal', 0.004893403687384717)
('basin_Lake Nyasa', 0.0038110333519065787)
('basin_Lake Rukwa', 0.0038094209883170445)
('basin_Lake Tanganyika', 0.003953663990236171)
('basin_Lake Victoria', 0.003878738630419606)
('basin_Pangani', 0.003048660084120809)
('basin_Rufiji', 0.0037036951332074263)
('basin_Ruvuma / Southern Coast', 0.003060860587954036)
('basin_Wami / Ruvu', 0.003347532750696452)
('region_Arusha', 0.0027126050982294618)
('region_Dar es Salaam', 0.0006458410595174205)
('region_Dodoma', 0.0016142415143666632)
('region_Iringa', 0.006021491999574731)
('region_Kagera', 0.0022687395386970454)
('region_Kigoma', 0.0027736886628620737

In [52]:
from sklearn.externals import joblib
filename = 'finalized_model.sav'
joblib.dump(clf, filename)

['finalized_model.sav']

In [53]:
df_test = pd.read_csv("Test data.csv")
df_test.head()

id  amount_tsh date_recorded                  funder  gps_height  \
0  50785         0.0    2013-02-04                    Dmdd        1996   
1  51630         0.0    2013-02-04  Government Of Tanzania        1569   
2  17168         0.0    2013-02-01                     NaN        1567   
3  45559         0.0    2013-01-22              Finn Water         267   
4  49871       500.0    2013-03-27                  Bruder        1260   

    installer  longitude   latitude                 wpt_name  num_private  \
0        DMDD  35.290799  -4.059696  Dinamu Secondary School            0   
1         DWE  36.656709  -3.309214                  Kimnyak            0   
2         NaN  34.767863  -5.004344           Puma Secondary            0   
3  FINN WATER  38.058046  -9.418672           Kwa Mzee Pange            0   
4      BRUDER  35.006123 -10.950412          Kwa Mzee Turuka            0   

           ...          payment_type water_quality quality_group  \
0          ...             never pay          soft          good   
1          ...             never pay          soft          good   
2          ...             never pay          soft          good   
3          ...               unknown          soft          good   
4          ...               monthly          soft          good   

       quantity  quantity_group                source           source_type  \
0      seasonal        seasonal  rainwater harvesting  rainwater harvesting   
1  insufficient    insufficient                spring                spring   
2  insufficient    insufficient  rainwater harvesting  rainwater harvesting   
3           dry             dry          shallow well          shallow well   
4        enough          enough                spring                spring   

   source_class     waterpoint_type waterpoint_type_group  
0       surface               other                 other  
1   groundwater  communal standpipe    communal standpipe  
2       surface               other                 other  
3   groundwater               other                 other  
4   groundwater  communal standpipe    communal standpipe  

[5 rows x 40 columns]

In [54]:
y_pred=clf.predict(X_test)

In [55]:
from sklearn.metrics import classification_report
print(classification_report(y_test, y_pred, target_names=["0", "1"]))

             precision    recall  f1-score   support

          0       0.81      0.85      0.83      8098
          1       0.81      0.76      0.79      6752

avg / total       0.81      0.81      0.81     14850

